internal/e2e/e2eapi/e2eapi_test.go

```diff
@@ -130,46 +117,71 @@ func TestDo(t *testing.T) {
 		defer ts.Close()
 
 		err := Do(ctx, http.MethodPost, ts.URL, nil, nil)
-		if err != nil {
-			t.Fatalf("exp nil err; got %v", err)
-		}
-	}()
-
-	for _, statusCode := range []int{http.StatusBadRequest, http.StatusOK} {
-		func() {
-			sentinel := errors.New("sentinel")
-			rtFn := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
-				res, err := http.DefaultClient.Do(req)
-				if err != nil {
-					return nil, err
-				}
-				res.Body = io.NopCloser(iotest.ErrReader(sentinel))
-				return res, nil
-			})
-
-			prev := defaultClient
-			defer func() {
-				defaultClient = prev
-			}()
-			defaultClient = new(http.Client)
-			defaultClient.Transport = rtFn
+		require.NoError(t, err)
+	})
+
+	// Covers IO errors
+	t.Run("IOError", func(t *testing.T) {
+
+		for _, statusCode := range []int{http.StatusBadRequest, http.StatusOK} {
+
+			// Covers IO errors for the sc >= 400 and default status code
+			// handling in the switch statement within do.
+			testName := fmt.Sprintf("Status=%v", http.StatusText(statusCode))
+			t.Run(testName, func(t *testing.T) {
+
+				// We assign a sentinel error to ensure propagation.
+				sentinel := errors.New("sentinel")
+
+				// This implementation of the http.RoundTripper is a way to
+				// cover the io.ReadAll(io.LimitReader(...)) lines in the switch
+				// statements inside do().
```

Comment:
The comments are much more helpful, yes.

I'm gonna approve, but just for future -- we should try to avoid specifying the "implementation" of the things we test. That's super fragile and if you change the code these tests either fail and you spend a lot of time figuring out why, or they become flaky -- wasting more time.

Reply:
I have a slightly different opinion around dependency injection and view the fact that changing code causes tests to fail as a benefit.

But I understand where you are coming from and will avoid this when I can.